core/src/main/java/kafka/server/share/SharePartition.java

```diff
@@ -3003,24 +3014,32 @@ private boolean hasOngoingStateTransition() {
          * if the state update is unsuccessful.
          *
          * @param newState The new state of the records.
-         * @param incrementDeliveryCount Whether to increment the delivery count.
+         * @param ops      The behavior on the delivery count.
          *
          * @return {@code InFlightState} if update succeeds, null otherwise. Returning state
          *         helps update chaining.
          */
-        private InFlightState tryUpdateState(RecordState newState, boolean incrementDeliveryCount, int maxDeliveryCount, String newMemberId) {
+        private InFlightState tryUpdateState(RecordState newState, DeliveryCountOps ops, int maxDeliveryCount, String newMemberId) {
+            boolean decreaseFlag = false;
             try {
+                if (ops == DeliveryCountOps.DECREASE && newState != RecordState.ARCHIVED) {
+                    deliveryCount--;
+                    decreaseFlag = true;
+                }
                 if (newState == RecordState.AVAILABLE && deliveryCount >= maxDeliveryCount) {
                     newState = RecordState.ARCHIVED;
                 }
                 state = state.validateTransition(newState);
-                if (incrementDeliveryCount && newState != RecordState.ARCHIVED) {
+                if (ops == DeliveryCountOps.INCREASE && newState != RecordState.ARCHIVED) {
                     deliveryCount++;
                 }
                 memberId = newMemberId;
                 return this;
             } catch (IllegalStateException e) {
                 log.error("Failed to update state of the records", e);
+                if (decreaseFlag) {
+                    deliveryCount++;
+                }
```

Comment:
We shouldn't need this change if we decrement the deliveryCount post stateTransition validation, correct?